query-algebrizer/src/clauses/where_fn.rs

```diff
@@ -42,28 +41,39 @@ use super::QualifiedAlias;
 
 use types::{
     ComputedTable,
+    EmptyBecause,
     SourceAlias,
     ValueTypeSet,
     VariableColumn,
 };
 
-macro_rules! check_type_set {
-    ($types: expr, $type: path) => { {
+macro_rules! coerce_to_typed_value {
+    ($var: ident, $val: ident, $types: expr, $type: path, $constructor: path) => { {
         if !$types.contains($type) {
-            bail!(ErrorKind::InvalidGroundConstant);
+            Impossible(EmptyBecause::TypeMismatch($var.clone(), $types, ValueTypeSet::of_one($type)))
+        } else {
+            Ok($constructor($val).into())
         }
     } }
 }
 
+enum ValueConversionResult {
```

Comment:
>  It "abuses" the `bail!` macro to return `ValueConversionResult::Err` instead of `std::result::Result::Err!` This surprised me greatly, and is (to my eye) a weakness in the Rust macro system.

Hah, it's a surprise to me, too! I hadn't even noticed that I'd left `bail!` in there, and that it was working.

I had originally implemented this with additional error cases, but that got really hairy: everywhere one might use `?` would end up as a `match` to turn some kinds of errors into `Ok` after processing.

I'll see if making this a `Result` is reasonable, and if not I'll take your other suggestion.